src/System.Collections.Sequences/System/Collections/Sequences/Position.cs

```diff
@@ -10,39 +10,59 @@ namespace System.Collections.Sequences
     public struct Position : IEquatable<Position>
     {
         long _index;
-        object _obj;
+        object _item;
 
         public static explicit operator int(Position position) => (int)position._index;
-        public static implicit operator long(Position position) => position._index;
-        public T As<T>() => _obj == null || IsInfinity ? default : (T)_obj;
-        public bool IsInfinity => this == Infinity;
 
-        public static readonly Position Infinity = new Position(new object(), long.MaxValue);
+        public static implicit operator long(Position position) => position._index;
 
-        public static Position operator +(Position position, long value) => new Position(position._obj, position._index + value);
+        public long GetIndex() => _index;
 
-        public static Position operator -(Position position, long value) => new Position(position._obj, position._index - value);
+        public T GetItem<T>() => _item == null || IsEnd ? default : (T)_item;
 
-        public static bool operator ==(Position left, Position right) => left._index == right._index && left._obj == right._obj;
-        public static bool operator !=(Position left, Position right) => left._index != right._index || left._obj != right._obj;
+        public (T item, long index) Get<T>() => (GetItem<T>(), this);
 
-        public void Set(long value) => _index = value;
+        public void SetIndex(long index) => _index = index;
 
-        public void Set(int value) => _index = value;
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public void SetItem<T>(T item) where T:class
+        {
+            if (item == null) this = End;
+            else _item = item;
+        }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public void Set<T>(T item)
+        public void Set<T>(T item, long index) where T : class
         {
-            if(EqualityComparer<T>.Default.Equals(item, default)) this = Infinity;
-            else _obj = item;
+            if (item == null) this = End;
+            else
+            {
+                _item = item;
+                _index = index;
+            }
         }
 
-        private Position(object obj, long index)
+        public void Advance<T>(T item, long offset) where T:class
         {
-            _obj = obj;
-            _index = index;
+            if (item == null) this = End;
+            else
+            {
+                _item = item;
+                _index += offset;
+            }
         }
 
+        public static Position operator +(Position position, long offset) => new Position(position._item, position._index + offset);
```

Comment:
Should this be checked to catch OverflowException?

Reply:
I am torn. I am not sure it adds enough value. If long index overflows, there will most certainly other check that will kick in. Also, if it overflows, it's just an int that then subsequently is to be used in data structures (e.g. arrays) that do checks. But I will think about it more.